### PR TITLE
Updates to GEDCOM parser

### PR DESCRIPTION
**Pull request by deeptowncitizen**
Interface extension:
1. Parsing of Maiden name
In case NAME with TYPE `maiden` is provided, it will be used to set maidenName
2. NCHI (number of children)
3. NMR (number of marriages)
4. images: an array of all images under the INDI, excluding the first (which is treated as the main image)
5. Image (string) data type was replaced with the custom structure (file + title). To avoid the breaking change, I added an additional method getMainImage()
6. Notes parser
7. General Events parser

### CODE REVIEW

File: src/gedcom.ts [hunk 1]
```diff
@@ -172,16 +192,30 @@ function createIndi(
...
+          : tag;
+          if (!realObjeTag) return null;
+
+          const file = findTag(realObjeTag.tree, 'FILE');
+          const title = findTag(realObjeTag.tree, 'TITL');
+
+          var result: any = {};
+          if (file) result.url = file.data;          
+          if (title) result.title = title.data;
+          return result;
```
Comment (PeWu) [inline src/gedcom.ts]:
```typescript
if (!file) {
  return undefined;
}
return ({
  url: file.data,
  title: title && title.data
});
```

File: src/data.ts [hunk 2]
```diff
@@ -56,13 +65,19 @@ export interface JsonGedcomData {
...
+  getMaidenName(): string | null;
   getBirthDate(): DateOrRange | null;
   getBirthPlace(): string | null;
   getDeathDate(): DateOrRange | null;
   getDeathPlace(): string | null;
   isConfirmedDeath(): boolean;
   getSex(): string | null;
   getImageUrl(): string | null;
+  getMainImage(): JsonImage | null;
+  getAlbumImages(): JsonImage[] | null;
```
Comment (PeWu) [inline src/data.ts]: `getImages()`

File: src/data.ts [hunk 3]
```diff
@@ -24,17 +24,26 @@ export interface JsonEvent extends DateOrRange {
   confirmed?: boolean;
 }
 
+export interface JsonImage {
+  url?: string;
```
Comment (PeWu) [inline src/data.ts]: url should not be optional

File: src/data.ts [hunk 4]
```diff
@@ -24,17 +24,26 @@ export interface JsonEvent extends DateOrRange {
...
+  maidenName?: string;
   famc?: string;
   fams?: string[];
+  numberOfChildren?: number;
+  numberOfMarriages?: number;
   birth?: JsonEvent;
   death?: JsonEvent;
   sex?: string;
-  imageUrl?: string;
+  mainImage?: JsonImage;
```
Comment (PeWu) [inline src/data.ts]: I'm not convinced a separate `mainImage` field is necessary. It should be enough to have one `images` field. We can always simply take the first element from the image array as the main image.

**Comment by deeptowncitizen** [comment]
@PeWu do you see any other issues?

File: src/gedcom.ts [hunk 5]
```diff
@@ -118,16 +133,21 @@ function createEvent(entry: GedcomEntry | undefined): JsonEvent | undefined {
...
+  const date = dateTag && dateTag.data && getDate(dateTag.data);
   const place = placeTag && placeTag.data;
   if (date || place) {
     const result: JsonEvent = date || {};
     if (place) {
       result.place = place;
     }
     result.confirmed = true;
+    result.type = typeTag ? typeTag!.data : undefined;
+    result.notes = notes;
```
Comment (PeWu) [inline src/gedcom.ts]:
```typescript
result.notes = createNotes(findTag(entry.tree, 'NOTE'));
```

File: src/gedcom.ts [hunk 6]
```diff
@@ -159,6 +186,25 @@ function createIndi(
     }
   }
 
+  if (maiden) {
+    const { firstName, lastName } = extractName(maiden.data);
```
Comment (PeWu) [inline src/gedcom.ts]:
```typescript
const { lastName } = extractName(maiden.data);
```

File: src/gedcom.ts [hunk 7]
```diff
@@ -148,9 +168,16 @@ function createIndi(
...
   // Name.
-  const nameTag = findTag(entry.tree, 'NAME');
-  if (nameTag) {
-    const { firstName, lastName } = extractName(nameTag.data);
+  const nameTags = findTags(entry.tree, 'NAME');
+  const isMaiden = (nameTag: GedcomEntry) => {
+    const type = findTag(nameTag.tree, 'TYPE');
+    return type !== undefined && type.data === 'maiden';
+  };
+  const main = nameTags.find(x => !isMaiden(x));
```
Comment (PeWu) [inline src/gedcom.ts]: If the only NAME tag provided is marked as *maiden*, we should also use it as the main name.
```typescript
const main = nameTags.length && (nameTags.find(x => !isMaiden(x)) || nameTags[0]);
```

File: src/gedcom.ts [hunk 8]
```diff
@@ -172,16 +218,26 @@ function createIndi(
...
+      if (!file) return undefined;
+      return {
+        url: file.data,
+        title: title && title.data,
+      } as JsonImage;
+    };
+
+    indi.images = objeTags
+      .map(getFileTag)
+      .filter((x): x is JsonImage => x !== undefined);
```
Comment (PeWu) [inline src/gedcom.ts]: This should work:
```typescript
.filter((x) => !!x);
```

Comment (deeptowncitizen) [reply src/gedcom.ts]: that's really ugly part. did not find a better option. eventually, this example does not work

File: src/gedcom.ts (hunk 7 above)
Comment (deeptowncitizen) [reply src/gedcom.ts]: I don't think it is correct to use maiden last name as the main. at least this is something that does not make sense to me.
1. user does not know the main lname, but knows maiden - we should not treat it as main as it adds an inconsistency
2. user added the main lname as maiden name... lib should not take responsibility for that imho
3. ???

File: src/gedcom.ts (hunk 5 above)
Comment (deeptowncitizen) [reply src/gedcom.ts]: original intent was to have a  single block of initialization. but as you prefer

**Comment by PeWu** [comment]
Thank you for the changes.

**Comment by PeWu** [comment]
I published version 2.5.0 with your changes.

**Comment by deeptowncitizen** [comment]
thank you

**Pull request merged**